plugins/linux-capture/pipewire.c

```diff
@@ -366,6 +375,152 @@ static void swap_texture_red_blue(gs_texture_t *texture)
 
+static inline struct spa_pod *build_format(struct spa_pod_builder *b,
+					   struct obs_video_info *ovi,
+					   uint32_t format, uint64_t *modifiers,
+					   size_t modifier_count)
+{
+	struct spa_pod_frame f[2];
+
+	/* make an object of type SPA_TYPE_OBJECT_Format and id SPA_PARAM_EnumFormat.
+	 * The object type is important because it defines the properties that are
+	 * acceptable. The id gives more context about what the object is meant to
+	 * contain. In this case we enumerate supported formats. */
+	spa_pod_builder_push_object(b, &f[0], SPA_TYPE_OBJECT_Format,
+				    SPA_PARAM_EnumFormat);
+	/* add media type and media subtype properties */
+	spa_pod_builder_add(b, SPA_FORMAT_mediaType,
+			    SPA_POD_Id(SPA_MEDIA_TYPE_video), 0);
+	spa_pod_builder_add(b, SPA_FORMAT_mediaSubtype,
+			    SPA_POD_Id(SPA_MEDIA_SUBTYPE_raw), 0);
+
+	/* formats */
+	spa_pod_builder_add(b, SPA_FORMAT_VIDEO_format, SPA_POD_Id(format), 0);
+
+	/* modifier */
+	if (modifier_count > 0) {
+		/* build an enumeration of modifiers */
+		spa_pod_builder_prop(b, SPA_FORMAT_VIDEO_modifier,
+				     SPA_POD_PROP_FLAG_MANDATORY |
+					     SPA_POD_PROP_FLAG_DONT_FIXATE);
+		spa_pod_builder_push_choice(b, &f[1], SPA_CHOICE_Enum, 0);
+		/* modifiers from  an array */
+		for (uint32_t i = 0; i < modifier_count; i++) {
+			uint64_t modifier = modifiers[i];
+			spa_pod_builder_long(b, modifier);
+			if (i == 0)
+				spa_pod_builder_long(b, modifier);
+		}
+		spa_pod_builder_pop(b, &f[1]);
+	}
+	/* add size and framerate ranges */
+	spa_pod_builder_add(b, SPA_FORMAT_VIDEO_size,
+			    SPA_POD_CHOICE_RANGE_Rectangle(
+				    &SPA_RECTANGLE(320, 240), // Arbitrary
+				    &SPA_RECTANGLE(1, 1),
+				    &SPA_RECTANGLE(8192, 4320)),
+			    SPA_FORMAT_VIDEO_framerate,
+			    SPA_POD_CHOICE_RANGE_Fraction(
+				    &SPA_FRACTION(ovi->fps_num, ovi->fps_den),
+				    &SPA_FRACTION(0, 1), &SPA_FRACTION(360, 1)),
+			    0);
+	return spa_pod_builder_pop(b, &f[0]);
+}
+
+static uint32_t build_format_params(obs_pipewire_data *obs_pw,
+				    struct spa_pod_builder *pod_builder,
+				    const struct spa_pod ***param_list)
+{
+	uint32_t n_params = 0;
+
+	const struct spa_pod **params;
+	params =
+		bzalloc(2 * obs_pw->format_info.num * sizeof(struct spa_pod *));
+
+	for (size_t i = 0; i < obs_pw->format_info.num; i++) {
+		if (obs_pw->format_info.array[i].modifiers.num == 0 ||
+		    !check_pw_version(&obs_pw->server_version, 0, 3, 33)) {
+			continue;
+		}
+		params[n_params++] = build_format(
+			pod_builder, &obs_pw->video_info,
+			obs_pw->format_info.array[i].spa_format,
+			obs_pw->format_info.array[i].modifiers.array,
+			obs_pw->format_info.array[i].modifiers.num);
+	}
+	for (size_t i = 0; i < obs_pw->format_info.num; i++) {
+		params[n_params++] = build_format(
+			pod_builder, &obs_pw->video_info,
+			obs_pw->format_info.array[i].spa_format, NULL, 0);
+	}
+	*param_list = params;
+	return n_params;
+}
+
+static void create_format_info(obs_pipewire_data *obs_pw)
```

Comment:
I appreciate that this acts on the `format_info` variable, but what this function effectively does is initialize the list of supported formats. So I'd rather call it `init_supported_formats` and `clear_supported_formats`.

Reply:
yeah makes sence, since the alloc stuff is gone with darray.